EDX-1719 Pick the Notifications MQTT/0MQ enhancement from v1.8 to v2.0

weichou1229 opened the pull request:
- Add fields to MQTTPubAddress
- Add ZeroMQPubAddress

FelixTing commented on dtos/address.go:
```diff
@@ -103,6 +117,20 @@ func NewEmailAddress(recipients []string) Address {
+	Topic string `json:"topic,omitempty" validate:"required"`
+}
+
+type ZeroMQPubAddress struct {
+}
+
+func NewZeroMQPubAddress(topic string) Address {
+	return Address{
+		Type:       common.EMAIL,
+		MessageBus: MessageBus{Topic: topic},
+	}
+}
```
this snippet seems incomplete

weichou1229 replied:
Change the wrong type from EMAIL to ZeroMQ.

weichou1229 replied:
The ZeroMQPubAddress DTO only need the topic field and Base Address's Type, Host, Port. So it is an empty struct to consist with other Address DTO.

FelixTing commented on models/address.go:
```diff
@@ -86,6 +92,13 @@ type MQTTPubAddress struct {
+	// SecretPath is used to specify the secret path to store the credential(username and password) for connecting the MQTT broker
+	// User need to store the credential via the /secret API before sending the notification
+	SecretPath string
+	// AuthMode is the SMTP authentication mechanism. Currently, 'usernamepassword' is the only AuthMode supported by this service, and the secret keys are 'username' and 'password'.
```
SMTP -> MQTT

In addition, for MQTT, I think the valid values for AuthMode should be 'usernamepassword', 'clientcert', 'cacert', and 'none'. See https://github.com/edgexfoundry/app-functions-sdk-go/blob/main/pkg/secure/mqttfactory.go#L87-L121 for reference.

weichou1229 replied:
Agree to have `clientcert`, but the `cacert` mode seems redundant.
```
case messaging.AuthModeCA:
		// Nothing to do here for this option
```

FelixTing replied:
Using `cacert` only is one of the approaches to establish an SSL/TLS connection.

The mosquito test server (https://test.mosquitto.org/, on port 8883) is a case in point.
The client only needs to verify the server certificate using the CA certificate available at https://test.mosquitto.org/ssl/mosquitto.org.crt.

weichou1229 replied:
I thought the user must provide the CA cert for TLS connection, so we don't need the AuthModeCA.
https://github.com/IOTechSystems/edgex-go-private/pull/244/files#diff-8f6d4a1a57476585a2a4b4dec737e7fcfcfd62fa71a8a096da68fa5c7fa3720bR83

FelixTing replied:
If I'm not mistaken, providing a CA certificate is not a must, it depends on whether the server certificate is signed by a trusted third party (i.e. a Certificate Authority). Generally, a set of trusted CA certificates are already included in client machines/softwares.

weichou1229 replied:
> whether the server certificate is signed by a trusted third party

Got it. 
- The user don't need to provide the CA cert if the cert is provided from trusted provider like letsencrypt. 
- the user need to provide the CA cert if the cert is self-signed .

FelixTing commented on dtos/address.go:
```diff
@@ -70,12 +82,15 @@ func NewRESTAddress(host string, port int, httpMethod string) Address {
+	Scheme         string `json:"scheme" validate:"required"`
+	SecretPath     string `json:"secretPath" validate:"required"`
+	AuthMode       string `json:"authMode" validate:"required,oneof='none' 'usernamepassword' 'cacert' 'clientcert'"`
+	SkipCertVerify bool   `json:"skipCertVerify"`
```
omitempty

weichou1229 replied:
Thanks for catch that. 
Added omitted to each fields because MQTTPubAddress is the part of Address DTO and we want to omit the fields for other struct like RESTAddress or EmailAddress in query result.

FelixTing reviewed (approved):
LGTM

cloudxxx8 reviewed (changes requested):
rename `ZeroMQPubAddress` to `ZeroMQAddress` 
move `Scheme` to `Address`
move the three security related fields to a new struct `Security`

weichou1229 commented:
> rename `ZeroMQPubAddress` to `ZeroMQAddress`
> move `Scheme` to `Address`
> move the three security related fields to a new struct `Security`

Renamed and modified the structs.

cloudxxx8 reviewed (approved):
LGTM
